### ISSUE TITLE
Inconsistent parsing behavior of next/router query parameters for Array values

### ISSUE THREAD

**Issue by matthewwolfe** [vercel/next.js#24673]
### What version of Next.js are you using?

10.2.0

### What version of Node.js are you using?

12.20.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start/next export

### Describe the Bug

The parsed output of `router.query` is inconsistent when array values are present as query parameters. Using the npm package `query-string` as a baseline, stepping through the various array format options produces unexpected behavior when parsing arrays.

Examples:

Url: `?example[]=test1`
Parsed: `{"example[]":"test1"}` (string value, brackets in property name not expected)

Url: `?example[]=test1&example[]=test2`
Parsed: `{"example[]":["test1","test2"]}` (array value, brackets in property name not expected)

Using `index`, `comma`, `separator`, and `bracket separator` formats do not produce an array ever

Using `none`, ie: `?example=test1&example=test2` is the only method that produces the correct output, although this is not ideal because single element arrays will be treated as a normal key/value string pair

### Expected Behavior

Parsed query behavior for arrays that is more aligned with the package `query-string`. If NextJS does not want to support the entirety of array formats, I would at least expect that the bracket format be updated to work completely, as it currently works in a partial and unexpected way:

Expected:
Url: `?example[]=test1`
Parsed: `{"example":["test1"]}` (Array value, no extra brackets in property name)

Url: `?example[]=test1&example[]=test2`
Parsed: `{"example":["test1","test2"]}` (Array value, no extra brackets in property name)

### To Reproduce

https://github.com/matthewwolfe/nextjs-router-query-array-example

**Comment by finfin** [comment]
I'm also confused by this.
After some study found that there's no clear querystring behavior spec defined, especially for array items.

Even if in query-string, the default behavior is the same with native querystring. 
Which is they treat single element as a string, and if same key appears more than once they will be combined as an array:
`?a=1` will be `{a: 1}`
`?a=1&a=2` will be turned to `{a: [1,2]}`
`?a[]=1` will be `{a[]: 1}`
`?a[]=1&a[]=2` will be `{a[]: [1,2]}`

Maybe another option is to let user opt in their preferred qs parsing lib?
Or use query-string as a base and give user more parsing options to choose?

**Comment by stefanprobst** [comment]
i guess eventually, query string parsing will move to `URLSearchParams` (see [here](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) and [here](https://nodejs.org/api/url.html#url_class_urlsearchparams)), so i'd be sceptical that `?example[]=` will be supported

**Comment by matthewwolfe** [comment]
@finfin I agree with you as well. We ended up writing a utility hook that uses `query-string` to parse strings more accurately, as the results of the built-in query param parsing is really wacky.

**Comment by JanKaifer** [comment]
Discussion is a more fitting place because it's not really a bug.
I'll move this there.

This would be a breaking change so we will probably not implement this unless we have a really good reason to introduce a breaking change.
